packages/client/src/devices/BrowserPermission.ts

```diff
@@ -0,0 +1,152 @@
+import { fromEventPattern, map } from 'rxjs';
+import { isReactNative } from '../helpers/platforms';
+import { getLogger } from '../logger';
+import { disposeOfMediaStream } from './devices';
+import { withoutConcurrency } from '../helpers/concurrency';
+
+interface BrowserPermissionConfig {
+  constraints: DisplayMediaStreamOptions;
+  queryName: PermissionName;
+}
+
+export class BrowserPermission {
+  private ready: Promise<void>;
+  private disposeController = new AbortController();
+  private state: PermissionState | undefined;
+  private wasPrompted: boolean = false;
+  private listeners = new Set<(state: PermissionState) => void>();
+  private logger = getLogger(['permissions']);
+
+  constructor(private readonly permission: BrowserPermissionConfig) {
+    const signal = this.disposeController.signal;
+
+    this.ready = (async () => {
+      const assumeGranted = (error?: unknown) => {
+        this.logger('warn', "Can't query permissions, assuming granted", {
+          permission,
+          error,
+        });
+        this.setState('granted');
+      };
+
+      if (!canQueryPermissions()) {
+        return assumeGranted();
+      }
+
+      try {
+        const status = await navigator.permissions.query({
+          name: permission.queryName,
+        });
+
+        if (!signal.aborted) {
+          this.setState(status.state);
+          status.addEventListener('change', () => this.setState(status.state), {
+            signal,
+          });
+        }
+      } catch (err) {
+        assumeGranted(err);
+      }
+    })();
+  }
+
+  dispose() {
+    this.state = undefined;
+    this.disposeController.abort();
+  }
+
+  async getState() {
+    await this.ready;
+    if (!this.state) {
+      throw new Error('BrowserPermission instance possibly disposed');
+    }
+    return this.state;
+  }
+
+  async prompt({
+    forcePrompt = false,
+    throwOnNotAllowed = false,
+  }: { forcePrompt?: boolean; throwOnNotAllowed?: boolean } = {}) {
+    await withoutConcurrency(
+      `permission-prompt-${this.permission.queryName}`,
+      async () => {
+        if (
+          (await this.getState()) !== 'prompt' ||
+          (this.wasPrompted && !forcePrompt)
```

Comment:
BrowserPermission will not allow re-prompting for a permission unless explicitly allowed to do so.